### ISSUE TITLE
Improve documentation on JSON Arrays Adapters

### ISSUE THREAD

**Issue by feinstein** [square/moshi#418]
The documentation shows how to parse a JSON Array Adapter as:
```
String cardsJsonResponse = ...;
Type type = Types.newParameterizedType(List.class, Card.class);
JsonAdapter<List<Card>> adapter = moshi.adapter(type);
List<Card> cards = adapter.fromJson(cardsJsonResponse);
```

This code uses a `Moshi` instance to call the `adapter()` method and generate an `Adapter`.

But when we look at the code examples on how to use `Moshi` with `Retrofit`, it starts to become a "Chicken-and-egg" problem. Here's the `Retrofit` code example:

```
Moshi moshi = new Moshi.Builder().add(new MyCustomAdapter()).build();
retrofit = new Retrofit.Builder()
                        .baseUrl(BASE_URL)
                        .addConverterFactory(MoshiConverterFactory.create(moshi))
                        .build();
```

So, I need to call the `Builder` `add` method to inject an `Adapter`, but since this is the `Builder`, there ins't a `Moshi` instance yet to create the JSON array adapter. Then how can I inject a `moshi` instance in a `MoshiConverterFactory.create` with the JSON array Adapter?

**Comment by JakeWharton** [comment]
You probably want a JsonAdapter.Factory into which the enclosing Moshi
instance is injected when someone asks for an adapter of a certain type.

**Comment by feinstein** [comment]
I see... I will try then to read the code and implement it this way, thanks Jake.

Jake, is it there any kind of caching of adapters? One behavior I am seeing is that if I create a JSON array Adapter, after Moshi is created, and never add this, just create, apparently, Retrofit will use it. I can provide a code example for this later today, I just left my laptop.

**Comment by aaronhe42** [comment]
Yes, adapters will be cached: https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/Moshi.java#L80

**Comment by feinstein** [comment]
Makes sense because I did the following code and it miraculously worked (kinda, I am still having issues parsing a `timestamp string` to a `ZonedDatetime`):
```
Type type = Types.newParameterizedType(List.class, Report.class);
Moshi moshi = new Moshi.Builder().build();
JsonAdapter<List<Report>> reportAdapter = moshi.adapter(type);
retrofit = new Retrofit.Builder()
                                   .baseUrl(BASE_URL)
                                   .addConverterFactory(MoshiConverterFactory.create(moshi))
                                   .build();
```

`reportAdapter` is never used, neither registered anywhere, so my guess was some sort of inner cache.

**Comment by feinstein** [comment]
@JakeWharton are there any examples or documentations about this? I am feeling kinda lost trying to use Moshi beyond the simple cases presented at the documentation.

I can't find any explanation on the purpose of the `JsonAdapter.Factory`, how it works with the rest of the framework, example of usages and etc.

Also, I made a simple adapter for dealing with timestamps and `ZonedDateTime` and it's trowing exceptions that don't make any sense to me, could someone please take a look at this stackoverflow case that I open for this issue:

**https://stackoverflow.com/questions/48143405/moshi-custom-adapter-isnt-being-called**

Thank you very much guys!

**Comment by feinstein** [comment]
@swankjesse what kind info are you missing?

**Comment by swankjesse** [comment]
Could you provide an executable example that throws exceptions that don't make sense? That's something we can take action on.

Last night I added a code sample for our JsonAdapter.Factory.  that might help also.

Finally I'd like to better understand what's Moshi needs from Retrofit. That's surprising.

**Comment by feinstein** [comment]
I will try to put here an executable, I am sorry, I thought my stackoverflow link above contained sufficient information.

I saw your code and I appreciate the example, but as I stated on #136, the usage of the factory is still a bit confusing to me, like when should I add an adapter and when  should I add an Adapter Factory.

Finally, I don't think my Moshi instance needs anything from Retrofit, my point of confusion was how to inject my moshi instance to Retrofit if the instance Builder needs a List Adapter and the documentations examples show that I need an instance, already created, in order to make the List adapter in the first place.

**Comment by NightlyNexus** [comment]
```java
Type type = Types.newParameterizedType(List.class, Report.class);
Moshi moshi = new Moshi.Builder().build();
JsonAdapter<List<Report>> reportAdapter = moshi.adapter(type);
retrofit = new Retrofit.Builder()
    .baseUrl(BASE_URL)
    .addConverterFactory(MoshiConverterFactory.create(moshi))
    .build();
```
What are you trying to accomplish in this code?
You don't need the instance of this adapter (`reportAdapter`) to give to anything. Retrofit's MoshiConverterFactory will call `moshi.adapter` and get this for you.

You only need to supply custom adapter factories in Moshi.Builder.add(...) if you are doing something custom with mapping your types. The adapter factories for lists, maps, primitives, non-platform Java types, etc. are already built in to the Moshi instance.

**Comment by feinstein** [comment]
@NightlyNexus yes, that's precisely what I figured earlier today and explained my source of confusion in #136 to Jake. 

In a nutshell, the documentation (the README.MD file) has a topic on `Parse JSON Arrays`, which led me to think I needed to provide one (and I forgot the beginning of the document states that Arrays Adapters are built-in). I honestly have no idea why the documentation has a topic on how to build an Adapter that's built-in already, AFAICS it should be removed or explained how it's different than the built-in Array Adapter.

**Comment by NightlyNexus** [comment]
It is the built-in adapter. The code in the readme shows how to get a list adapter from Moshi (which Moshi will either create or pull from the internal cache).

The user of the library is not writing the code for the adapter or its factory, merely getting it from Moshi in order to use it to serialize a JSON array.

**Comment by feinstein** [comment]
I don't have this code at all on my project and I am able to deserialize a JSON array. Are you saying every time I want to serialize a List to a JSON Array I will

**Comment by feinstein** [comment]
... Reading the documentation it's clearly deserializing the JSON Array, so I still don't get it why I have to create that Adapter, as the documentations states, if my code works perfectly without it (it reads a List<Report> from JSON).

No offenses, but I feel like I am more memorizing how to make Moshi work, than to understand how it works. It would be nice of you guys could put up together some sort of document, explaining how it works internally, so people like me could actually understand what to give Moshi and when, specially on unusual use cases.

**Comment by JakeWharton** [comment]
You don't have to. As far as I can tell you should be doing `MoshiConverterFactory.create()`, passing to Retrofit, and that's it.

**Comment by feinstein** [comment]
... So I did some thinking... , I am using Moshi directly with Retrofit... So are you saying that Retrofit is using the code in `Parsing a JSON Array` internally to get my List for me and that's why it works in my case, where I don't write any of that code?

**Comment by NightlyNexus** [comment]
>[Retrofit's MoshiConverterFactory will call moshi.adapter and get this for you.](https://github.com/square/retrofit/blob/c2b44bd3b63cbbdc4c56014b13e463b72a23ec8d/retrofit-converters/moshi/src/main/java/retrofit2/converter/moshi/MoshiConverterFactory.java#L91)

Yes.

**Comment by mabuthraa** [comment]
It's not different issue .. it's actually that MoshiConverterFactory doesnt parse array.. I got this : Unable to create converter for name of DTO ...

my dto:
data class GetFeedResponseDto(
    val currentItemCount: Int,
    val items: List<MessageItemDto>,
    val totalPages: Int
)
